source/extensions/load_balancing_policies/client_side_weighted_round_robin/client_side_weighted_round_robin_lb.cc

```diff
@@ -0,0 +1,335 @@
+#include "source/extensions/load_balancing_policies/client_side_weighted_round_robin/client_side_weighted_round_robin_lb.h"
+
+#include <sys/types.h>
+
+#include <cstdint>
+#include <optional>
+#include <string>
+
+#include "envoy/common/time.h"
+#include "envoy/upstream/upstream.h"
+
+#include "source/common/orca/orca_load_metrics.h"
+#include "source/common/protobuf/utility.h"
+#include "source/extensions/load_balancing_policies/common/load_balancer_impl.h"
+
+#include "absl/status/status.h"
+#include "xds/data/orca/v3/orca_load_report.pb.h"
+
+using Envoy::MonotonicTime;
+using ::envoy::extensions::load_balancing_policies::client_side_weighted_round_robin::v3::
+    ClientSideWeightedRoundRobin;
+using Envoy::Upstream::Host;
+using xds::data::orca::v3::OrcaLoadReport;
+
+#if TEST_THREAD_SUPPORTED
+#define IS_MAIN_OR_TEST_THREAD() (Envoy::Thread::MainThread::isMainOrTestThread())
+#else // !TEST_THREAD_SUPPORTED -- just check for the main thread.
+#define IS_MAIN_OR_TEST_THREAD() (Envoy::Thread::MainThread::isMainThread())
+#endif // TEST_THREAD_SUPPORTED
+
+namespace {
+
+std::string getHostAddress(const Host* host) {
+  if (host == nullptr || host->address() == nullptr) {
+    return "unknown";
+  }
+  return host->address()->asString();
+}
+
+} // namespace
+
+namespace Envoy {
+namespace Upstream {
+
+ClientSideWeightedRoundRobinLoadBalancer::ClientSideWeightedRoundRobinLoadBalancer(
+    const PrioritySet& priority_set, const PrioritySet* local_priority_set, ClusterLbStats& stats,
+    Runtime::Loader& runtime, Random::RandomGenerator& random,
+    const envoy::config::cluster::v3::Cluster::CommonLbConfig& common_config,
+    const ClientSideWeightedRoundRobin& client_side_weighted_round_robin_config,
+    TimeSource& time_source, Event::Dispatcher& main_thread_dispatcher)
+    : EdfLoadBalancerBase(
+          priority_set, local_priority_set, stats, runtime, random,
+          PROTOBUF_PERCENT_TO_ROUNDED_INTEGER_OR_DEFAULT(common_config, healthy_panic_threshold,
+                                                         100, 50),
+          LoadBalancerConfigHelper::localityLbConfigFromCommonLbConfig(common_config),
+          /*slow_start_config=*/std::nullopt, time_source) {
+  ENVOY_LOG(trace, "RoundRobinLbConfig config {}",
+            client_side_weighted_round_robin_config.DebugString());
+  initFromConfig(client_side_weighted_round_robin_config);
+  if (IS_MAIN_OR_TEST_THREAD()) {
+    startWeightUpdatesOnMainThread(main_thread_dispatcher);
+  }
```

Comment:
It's weird to do this in this way. Note, all LBs now is wrapped as thread aware lb. If you want to do something periodically at main thread. Just set it in the `initialize()` method of thread aware lb.

The `Common::FactoryBase` is helper to simplify the code of the LBs who needn't main thread state. But the new LB may couldn't use it because may need to do something at main thread.

You may need to do following things in the initialize():

1. set client side policy data to all hosts;
2. set client side policy data to new hosts when hosts are added;
3. set a timer to refresh weight periodically.

Reply:
I've done (1) and (3), but I've missed (2). I'll update it shortly.

Reply:
Done.